File: packages/react-native-carplay/ios/RNCarPlayViewController.m
//
//  RNCarPlayViewController.m
//  react-native-carplay
//
//  Created by Susan Thapa on 27/02/2024.
//

#import "RNCarPlayViewController.h"
#import <Foundation/Foundation.h>
#import <React/RCTRootView.h>

@interface RNCarPlayViewController ()

@property(nonatomic, strong) RCTRootView *rootView;

@end

@implementation RNCarPlayViewController

- (instancetype)initWithRootView:(RCTRootView *)rootView {
    self = [super init];
    if (self) {
        _rootView = rootView;
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    self.view.translatesAutoresizingMaskIntoConstraints = false;
    if (self.rootView) {
        self.rootView.translatesAutoresizingMaskIntoConstraints = false;
        self.rootView.frame = self.view.bounds;
        [self.view addSubview:self.rootView];

        [NSLayoutConstraint activateConstraints:@[
            [self.rootView.leadingAnchor
                constraintEqualToAnchor:self.view.leadingAnchor],
            [self.rootView.trailingAnchor
                constraintEqualToAnchor:self.view.trailingAnchor],
            [self.rootView.topAnchor
                constraintEqualToAnchor:self.view.topAnchor],
            [self.rootView.bottomAnchor
                constraintEqualToAnchor:self.view.bottomAnchor],
        ]];
    }
}

- (void)viewWillLayoutSubviews {
    [super viewWillLayoutSubviews];
    self.rootView.frame = self.view.bounds;
}

- (void)viewDidLayoutSubviews {
    [super viewDidLayoutSubviews];

    UIEdgeInsets currentSafeAreaInsets = self.view.safeAreaInsets;
    sendRNCarPlayEvent(@"safeAreaInsetsChanged", @{
        @"bottom" : @(currentSafeAreaInsets.bottom),
        @"left" : @(currentSafeAreaInsets.left),
        @"right" : @(currentSafeAreaInsets.right),
        @"top" : @(currentSafeAreaInsets.top),
        @"templateId" : self.rootView.moduleName
    });
}

@end

File: packages/react-native-carplay/ios/RNCarPlayViewController.swift
//
//  RNCarPlayViewController.swift
//  RNCarPlay
//
//  Created by Susan Thapa on 27/02/2024.
//  Updated by Manuel Auer on 24.10.24.
//

import React

@objc(RNCarPlayViewController)
public class RNCarPlayViewController: UIViewController {
    let rootView: RCTRootView
    
    @objc public init(rootView: RCTRootView) {
        self.rootView = rootView
        super.init(nibName: nil, bundle: nil)
    }

    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    public override func viewDidLoad() {
        super.viewDidLoad()

        self.view.translatesAutoresizingMaskIntoConstraints = false
        self.rootView.frame = self.view.bounds
        self.view = rootView
    }

    public override func viewWillLayoutSubviews() {
        super.viewWillLayoutSubviews()
        self.rootView.frame = self.view.bounds
    }

    public override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()

        RNCarPlayUtils.sendRNCarPlayEvent(
            name: "safeAreaInsetsDidChange",
            body: [
                "bottom": self.view.safeAreaInsets.bottom,
                "left": self.view.safeAreaInsets.left,
                "right": self.view.safeAreaInsets.right,
                "top": self.view.safeAreaInsets.top,
                "id": self.rootView.moduleName,
            ])
    }

    public override func traitCollectionDidChange(
        _ previousTraitCollection: UITraitCollection?
    ) {
        super.traitCollectionDidChange(previousTraitCollection)
        if traitCollection.hasDifferentColorAppearance(
            comparedTo: previousTraitCollection)
        {
            RNCarPlayUtils.sendRNCarPlayEvent(
                name: "appearanceDidChange",
                body: [
                    "colorScheme": traitCollection.userInterfaceStyle == .dark
                        ? "dark" : "light",
                    "id": self.rootView.moduleName,
                ])
        }
    }
    
    deinit {
        self.view = nil
    }
}
